File: c/cap5/cap5.part1ex04.c
/* Copyright (C) 2022 Francisco de Assis Zampirolli - All Rights Reserved
 * Author: Francisco de Assis Zampirolli
 * Purpose: Examples and some exercises from the Information Processing book
 * Language: C
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>5.
 */
#include <stdio.h>
#include <stdlib.h>  // malloc e free

int* leiaVetor(int n) {
  int* v = malloc(n * sizeof(int));
  for (int i = 0; i < n; i++) {
    scanf("%d", &v[i]);
  }
  return v;
}

void escrevaVetor(int* v, int n) {
  for (int i = 0; i < n; i++) {
    printf("%d\n", v[i]);
  }
}

int main(void) {
  // ENTRADA DE DADOS
  int n, * v1;   // variaveis de referência v1
  printf("Digite o tamanho do vetor: ");
  scanf("%d", &n);
  int* v2 = malloc(100 * sizeof(int));
  printf("Digite os elementos: ");
  v1 = leiaVetor(n);
  // PROCESSAMENTO
  for (int i = 0; i < n; i++) {
    int max = v1[i];
    if (i - 1 >= 0 && max < v1[i - 1])
      max = v1[i - 1];
    if (i + 1 < n && max < v1[i + 1])
      max = v1[i + 1];
    v2[i] = max;
  }
  // SAÍDA:
  printf("\nv2:\n");
  escrevaVetor(v2, n);
  free(v1); // liberar memória alocado com malloc
  free(v2);
  return 0;
}

File: py/cap5/cap5.part1ex04.py
"""
Copyright(C) 2022 - 2023 Francisco de Assis Zampirolli - All Rights Reserved
Purpose : Examples and some exercises from the Information Processing book
Language : Python
"""

import numpy as np  # muito útil para trabalhar com vetor!


# MÉTODOS
def leiaVetor(n):
    v = np.zeros(n).astype(int)  # aloca vetor de inteiro com n elementos
    for i in range(n):
        v[i] = int(input())
    return v


def escrevaVetor(v):
    for i in range(n):
        print(v[i])


# ENTRADA DE DADOS
n = int(input("Digite o tamanho do vetor:"))
print("Entre com " + str(n) + " elementos:")
v1 = leiaVetor(n)

v2 = np.zeros(n).astype(int)

# PROCESSAMENTO
for i in range(n):
    max = v1[i]
    if i - 1 >= 0 and max < v1[i - 1]:
        max = v1[i - 1]
    if i + 1 < n and max < v1[i + 1]:
        max = v1[i + 1]
    v2[i] = max

# SAÍDA DE DADOS
print("v2:")
escrevaVetor(v2)
